### ISSUE TITLE
Win 10 Insiders builds changes path from \\wsl$\ to \\wsl.localhost\ and breaks docker

### ISSUE THREAD

**Issue by jflam** [docker/for-win#10786]
- [ x] I have tried with the latest version of Docker Desktop
  - [ x] I have tried disabling enabled experimental features
  - [ ] I have uploaded Diagnostics
  - Diagnostics ID:

### Actual behavior

Docker CLI in Windows is unhappy with `\\wsl.localhost\...` as a path in the command below that is run using `cmd.exe`. Recent Windows 10 Insider builds changed the format of paths to WSL2 directories from `\\wsl\localhost\...` to `\\wsl.localhost\...`: 

```
docker run --sig-proxy=false -a STDOUT -a STDERR --mount type=bind,source=\\wsl.localhost\Ubuntu-18.04\home\jlam\src\ez\src\pytorch-tutorials,target=/workspaces/pytorch-tutorials,consistency=cached --mount source=/var/run/docker.sock,target=/var/run/docker.sock,type=bind --mount type=volume,src=vscode,dst=/vscode -l vsch.local.folder=\\wsl.localhost\Ubuntu-18.04\home\jlam\src\ez\src\pytorch-tutorials -l vsch.quality=stable -l vsch.remote.devPort=0 -u ezuser --gpus=all --ipc=host --entrypoint /bin/sh pytorch-tutorials -c echo Container started ; trap "exit 0" 15; while sleep 1 & wait $!; do :; done
docker: Error response from daemon: \\wsl.localhost\Ubuntu-18.04\home\jlam\src\ez\src\pytorch-tutorials%!(EXTRA string=is not a valid Windows path).
See 'docker run --help'.
'wait' is not recognized as an internal or external command,
operable program or batch file.
```

### Expected behavior

docker run runs the container.

### Information

  - Windows Version: Win 10 Insiders 21349.1000
  - Docker Desktop Version: Docker version 20.10.5, build 55c4c88
  - WSL2 or Hyper-V backend?  WSL 2
  - Are you running inside a virtualized Windows e.g. on a cloud server or a VM: No

### Steps to reproduce the behavior

  1. You will need a recent Win10 Insiders Build. I'm using `Win 10 Insiders 21349.1000`
  2. If you try to run a container using a path to a directory in WSL 2, e.g., `\\wsl.localhost\Ubuntu-18.04\home\jlam\src\ez\src\pytorch-tutorials` in my example above, docker run will fail.

It looks like docker run is unhappy with the format of the path, in particular the change to `\\wsl.localhost\` from from `\\wsl\localhost` from earlier Windows Insider builds.

**Comment by stephen-turner** [comment]
Thanks for the heads-up. I saw your ticket on microsoft/vscode-remote-release, but do you know if there's any ticket at microsoft/WSL, or any documentation about the change? Before changing anything on our side, I'd like to know whether the change is intentional and permanent, and whether there are any other pathnames we might have to parse.

**Comment by jflam** [comment]
I pinged some folks I know in Windows to see if they can dig something official up. Stay tuned. Thx!

**Comment by stephen-turner** [comment]
Thanks, we've reached out to our contacts in Microsoft too. We will support the new path.

**Comment by jflam** [comment]
Excellent! It looks like I was the first person to run into this as that build hasn't hit any of the public fast rings yet.

**Comment by mauve** [comment]
I am on Insider Preview 21354 and this just broke everything, is there a workaround?

**Comment by stephen-turner** [comment]
Well, you can still use the \\wsl$\ syntax.

**Comment by mauve** [comment]
@stephen-turner hmm, I am currently using it via VSCode, do you know if there is a way to get VSCode to use the old syntax?

**Comment by stephen-turner** [comment]
That I don't know, I'm afraid.

**Comment by andyli** [comment]
FYI, there is a paragraph in https://blogs.windows.com/windows-insider/2021/04/07/announcing-windows-10-insider-preview-build-21354/:

> The default path to access Linux files from Windows using the Windows Subsystem for Linux has been changed from `\\wsl\\` to `\\wsl.localhost\\` to improve performance and reliability. You will still be able to access Linux files using `\\wsl$\\`.

**Comment by StefanScherer** [comment]
We're working on a fix that should be part of the next patch release 3.3.1.

**Comment by StefanScherer** [comment]
We have a fix in a preview build that I can share with you: https://desktop-stage.docker.com/win/stable/amd64/63026/Docker%20Desktop%20Installer.exe
We would love to head feedback if that fully fixes your problem.

**Comment by jflam** [comment]
Can confirm this works on 21354.1. Thanks for the quick turnaround!

**Comment by mauve** [comment]
can confirm it works on 21359, thanks for the quick fix

**Comment by StefanScherer** [comment]
The issue is fixed with Docker Desktop 3.3.1 release.
